File: vc-verifier/kotlin/vcverifier/src/main/java/io/mosip/vercred/vcverifier/credentialverifier/validator/LdpValidator.kt
package io.mosip.vercred.vcverifier.credentialverifier.validator

import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.CONTEXT
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.CREDENTIAL_SCHEMA
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.CREDENTIAL_STATUS
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.CREDENTIAL_SUBJECT
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_CODE_EMPTY_VC_JSON
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_CODE_GENERIC
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_CODE_INVALID
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_CODE_MISSING
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_CODE_VC_EXPIRED
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_MESSAGE_CONTEXT_FIRST_LINE
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_MESSAGE_EMPTY_VC_JSON
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_MESSAGE_VC_EXPIRED
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ERROR_MISSING_REQUIRED_FIELDS
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.EVIDENCE
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.EXCEPTION_DURING_VALIDATION
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.EXPIRATION_DATE
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ISSUANCE_DATE
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.ISSUER
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.PROOF
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.REFRESH_SERVICE
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.TERMS_OF_USE
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.TYPE
import io.mosip.vercred.vcverifier.constants.CredentialValidatorConstants.VALID_UNTIL
import io.mosip.vercred.vcverifier.data.DATA_MODEL
import io.mosip.vercred.vcverifier.data.ValidationStatus
import io.mosip.vercred.vcverifier.exception.ValidationException
import io.mosip.vercred.vcverifier.utils.DateUtils
import io.mosip.vercred.vcverifier.utils.Util
import io.mosip.vercred.vcverifier.utils.ValidationHelper
import org.json.JSONObject

class LdpValidator {

    private val commonMandatoryFields = listOf(
        CONTEXT,
        TYPE,
        CREDENTIAL_SUBJECT,
        ISSUER,
        PROOF
    )

    //All Fields has Type Property as mandatory with few fields as ID as mandatory.
    private val allFieldsWithIDAndType = listOf(
        PROOF,
        CREDENTIAL_STATUS,
        EVIDENCE,
        CREDENTIAL_SCHEMA,
        REFRESH_SERVICE,
        TERMS_OF_USE
    )

    private val validationHelper = ValidationHelper()
    private val dateUtils = DateUtils

    fun validate(credential: String): ValidationStatus {
        try {
            if (credential.isEmpty()) {
                throw ValidationException(ERROR_MESSAGE_EMPTY_VC_JSON, ERROR_CODE_EMPTY_VC_JSON)
            }

            val vcJsonObject = JSONObject(credential)

            val contextVersion = Util.getContextVersion(vcJsonObject)
                ?: throw ValidationException("$ERROR_MISSING_REQUIRED_FIELDS$CONTEXT", "${ERROR_CODE_MISSING}${CONTEXT.uppercase()}")
            when (contextVersion) {
                DATA_MODEL.DATA_MODEL_1_1 -> {
                    validateV1SpecificFields(vcJsonObject)
                    validateCommonFields(vcJsonObject)
                    val expirationMessage = if (vcJsonObject.has(EXPIRATION_DATE) && dateUtils.isVCExpired(vcJsonObject.optString(
                            EXPIRATION_DATE))) ERROR_MESSAGE_VC_EXPIRED else ""
                    val verificationStatusCode = if (vcJsonObject.has(EXPIRATION_DATE) && dateUtils.isVCExpired(vcJsonObject.optString(
                            EXPIRATION_DATE))) ERROR_CODE_VC_EXPIRED else ""
                    return ValidationStatus(expirationMessage, verificationStatusCode)
                }
                DATA_MODEL.DATA_MODEL_2_0 -> {
                    validateV2SpecificFields(vcJsonObject)
                    validateCommonFields(vcJsonObject)
                    val expirationMessage = if (vcJsonObject.has(VALID_UNTIL) && dateUtils.isVCExpired(vcJsonObject.optString(VALID_UNTIL))) ERROR_MESSAGE_VC_EXPIRED else ""
                    val verificationStatusCode = if (vcJsonObject.has(VALID_UNTIL) && dateUtils.isVCExpired(vcJsonObject.optString(
                            VALID_UNTIL))) ERROR_CODE_VC_EXPIRED else ""
                    return ValidationStatus(expirationMessage, verificationStatusCode)
                }
                else -> {
                    throw ValidationException(ERROR_MESSAGE_CONTEXT_FIRST_LINE, "${ERROR_CODE_INVALID}${CONTEXT.uppercase()}")
                }
            }

        }
        catch (e: ValidationException) {
            return ValidationStatus(e.errorMessage, e.errorCode)
        }
        catch (e: Exception) {
            return ValidationStatus("${EXCEPTION_DURING_VALIDATION}${e.message}", ERROR_CODE_GENERIC)
        }

    }

    //Validation for Data Model 1.1
    private fun validateV1SpecificFields(vcJsonObject: JSONObject) {

        val v1SpecificMandatoryFields = listOf(
            ISSUANCE_DATE
        )

        validationHelper.checkMandatoryFields(vcJsonObject, commonMandatoryFields+v1SpecificMandatoryFields)

        dateUtils.validateV1DateFields(vcJsonObject)

        val v1SpecificIdMandatoryFields = listOf(
            CREDENTIAL_STATUS,
            REFRESH_SERVICE,
            CREDENTIAL_SCHEMA
        )

        allFieldsWithIDAndType.forEach { field ->
            if(vcJsonObject.has(field)){
                validationHelper.validateFieldsByIdAndType(
                    vcJsonObject = vcJsonObject,
                    fieldName = field,
                    idMandatoryFields = v1SpecificIdMandatoryFields
                )
            }
        }

    }

    //Validation for Data Model 2.0
    private fun validateV2SpecificFields(vcJsonObject: JSONObject){

        validationHelper.checkMandatoryFields(vcJsonObject, commonMandatoryFields)

        dateUtils.validateV2DateFields(vcJsonObject)

        val v2SpecificIdMandatoryFields = listOf(
            CREDENTIAL_SCHEMA
        )

        allFieldsWithIDAndType.forEach { field ->
            if(vcJsonObject.has(field)){
                validationHelper.validateFieldsByIdAndType(vcJsonObject = vcJsonObject,
                    fieldName = field,
                    idMandatoryFields = v2SpecificIdMandatoryFields
                )
            }
        }

        validationHelper.validateNameAndDescription(vcJsonObject)


    }

    //Common Validations
    private fun validateCommonFields(vcJsonObject: JSONObject){

        validationHelper.validateCredentialSubject(vcJsonObject)

        validationHelper.validateProof(vcJsonObject.toString())

        validationHelper.validateId(vcJsonObject)

        validationHelper.validateType(vcJsonObject)

        validationHelper.validateIssuer(vcJsonObject)

    }
}

File: vc-verifier/js/src/validator/LdpValidator.js
import {Errors, Fields} from "../constant/ValidatorConstants.js";
import {
    DataModel,
    getContextVersion,
    validateCredentialSubject, validateFieldsWithID, validateFieldsWithType, validateID, validateIssuer,
    validateMandatoryFields, validateNameAndDescription, validateProof, validateType, verificationFailure, verificationSuccess
} from "./ValidationHelper.js";
import {ValidationError} from "./ValidationError.js";
import {isVCExpired, validateV1SpecificDateFields, validateV2SpecificDateFields} from "./DateUtils.js";
import {isEmptyStringOrEmptyObject} from "./Utils.js";


const commonMandatoryFields = [Fields.CONTEXT, Fields.TYPE, Fields.CREDENTIAL_SUBJECT, Fields.PROOF, Fields.ISSUER]

export const validate = (credential) => {
    try {
        if(isEmptyStringOrEmptyObject(credential)) {
            throw new ValidationError(
                Errors.ERROR_EMPTY_VC_JSON);
        }

        let contextVersion = getContextVersion(credential)

        switch (contextVersion){
            case DataModel.DATA_MODEL_1_1 : {
                v1SpecificFieldsValidation(credential)
                commonValidations(credential)
                const expirationMessage = (credential.hasOwnProperty(Fields.EXPIRATION_DATE) && isVCExpired(credential[Fields.EXPIRATION_DATE]))
                    ? Errors.ERROR_VC_EXPIRED
                    : "";
                return verificationSuccess(expirationMessage)
            }
            case DataModel.DATA_MODEL_2_0: {
                v2SpecificFieldsValidation(credential)
                commonValidations(credential)
                const expirationMessage = (credential.hasOwnProperty(Fields.VALID_UNTIL) && isVCExpired(credential[Fields.VALID_UNTIL]))
                    ? Errors.ERROR_VC_EXPIRED
                    : "";
                return verificationSuccess(expirationMessage)
            }
        }

    } catch (error){
        if(error instanceof ValidationError){
            return verificationFailure(`${error.message}`)
        } else {
            return verificationFailure(`${Errors.EXCEPTION_DURING_VALIDATION}${error.message}`)
        }

    }

}

const v1SpecificFieldsValidation = (credential) => {
    const v1specificMandatoryFields = [...commonMandatoryFields, Fields.ISSUANCE_DATE];
    validateMandatoryFields(credential, v1specificMandatoryFields);
    validateV1SpecificDateFields(credential);

    const allFieldsWithTypeAndId = [
        Fields.CREDENTIAL_STATUS,
        Fields.EVIDENCE,
        Fields.TERMS_OF_USE,
        Fields.REFRESH_SERVICE,
        Fields.CREDENTIAL_SCHEMA,
        Fields.PROOF
    ];

    validateFieldsWithType(credential, allFieldsWithTypeAndId);
    validateFieldsWithID(credential, allFieldsWithTypeAndId, [
        Fields.CREDENTIAL_STATUS,
        Fields.REFRESH_SERVICE,
        Fields.CREDENTIAL_SCHEMA
    ]);
};

const v2SpecificFieldsValidation = (credential) => {
    const v2specificMandatoryFields = [ ...commonMandatoryFields]
    validateMandatoryFields(credential, v2specificMandatoryFields)

    validateV2SpecificDateFields(credential)

    validateNameAndDescription(credential)

    const allFieldsWithTypeAndId = [
        Fields.CREDENTIAL_STATUS,
        Fields.EVIDENCE,
        Fields.TERMS_OF_USE,
        Fields.REFRESH_SERVICE,
        Fields.CREDENTIAL_SCHEMA,
        Fields.PROOF
    ];
    validateFieldsWithType(credential, allFieldsWithTypeAndId)

    const v2IDMandatoryFields = [
        Fields.CREDENTIAL_SCHEMA
    ]
    validateFieldsWithID(credential, allFieldsWithTypeAndId, v2IDMandatoryFields)
}

const commonValidations = (credential) => {

    validateType(credential)

    validateID(credential)

    validateIssuer(credential)

    validateProof(credential)

    validateCredentialSubject(credential)


}





